### ISSUE TITLE
offline-plugin and updates

### ISSUE THREAD

**Issue by NekR** [react-boilerplate/react-boilerplate#1113]
## Issue Type
- [x] Feature-Quetion-Problem
## The problem

The problem is that many people don't get how `ServiceWorker` works and think that `offline-plugin` messes up with their builds, they build to the production and nothing changes, so they think imports isn't working or build itself works incorrectly. This is based on real facts and I even got hired once to help to deal with `offline-plugin` because client thought it messed up his builds.
## The solution

There is no silver bullet. But I have couple of suggestions:
- Built-in UI toast / tooltip which is displayed when new version is available and says "Click here to update"
- Automatically apply update once it's ready and **auto reload all open pages**. This would look like this:

``` js
const runtime = require('offline-plugin/runtime');

runtime.install({
  onUpdating: () => {
    console.log('SW Event:', 'onUpdating');
  },
  onUpdateReady: () => {
    console.log('SW Event:', 'onUpdateReady');
    // Update is applied here
    runtime.applyUpdate();
  },
  onUpdated: () => {
    console.log('SW Event:', 'onUpdated');
    // Force reload happens here
    window.location.reload();
  },

  onUpdateFailed: () => {
    console.log('SW Event:', 'onUpdateFailed');
  }
});
```

Both of this solution are a bit weird and unfortunate, but I don't see any other ways right. Ideas?

**Comment by mxstbr** [comment]
What if we combine the two? Automatically apply the update, and once it's ready show a UI tooltip which, when clicked, reloads the page? That way avoid auto reloads, but the user also doesn't need to do anything!

**Comment by NekR** [comment]
Yeah, that is kind of what I meant for first way. Except I thought to apply
update once user clicks on a toast. There is a possible problem in applying
an update before knowing for sure that we are going to reload the page. If
user doesn't reload the page in any way, then there could be conflicts
between project version.

Maybe we need this not for a production but for a dev environment?
Currently `offline-plugin` is disabled in dev because there was no sense to
run it with dev-server. Now there might be one -- teach devs how to deal
with `offline-plugin` and SW in general.

I can try to make `offline-plugin` work in dev-server as it works on
production. So we can teach devs about it even after first change in a
project.

I'm still a bit sure though, it's a hard topic :-(

**Comment by mxstbr** [comment]
No I like it, I wouldn't enable updates-only-on-click in dev, but in general I think this would be great!

**Comment by NekR** [comment]
@mxstbr Just so we are clear, can you summarize what is good / what you like?

**Comment by mxstbr** [comment]
Sorry, in production we should show a toast which is clickable to update to a new version!

**Comment by NekR** [comment]
Okay, I'll try do a PR this week or so.

**Comment by NekR** [comment]
@mxstbr hey.. I kinda forgot about it, sorry. But here is a good news: I released v4 of `offline-plugin` and now waiting to use it for some time. Once I'm sure it's stable enough I'll be updating it in `react-boilerplate` and will do this issue in the same time.

Max, can you assign it to me so I could easily track it (if it's possible)?

**Comment by gihrig** [comment]
@NekR 
>Once I'm sure it's stable enough I'll be updating it in react-boilerplate

I plan on submitting dependency update this weekend. Should I hold off moving to 4.5.x?

PS. I assigned this issue to you. Looking forward to your work 😄

**Comment by NekR** [comment]
@gihrig
> I plan on submitting a dependency update this weekend. Should I hold off moving to 4.5.x?

That's really hard to tell. v4 has been there for a while (in a separate branch / npm tag) and merged it to master just a few days ago. I recently built a PWA with 4.5.* and fixed few bugs and then it seemed to work as expected. Though, I certainly didn't expect those bugs to be there so there might be something else too.

What I meant here is that I don't want to break things for people who use `react-boilerplate` and that there isn't enough feedback from people about v4 yet. So maybe it indeed makes sense to hold off with moving to v4 right now.

**Comment by gihrig** [comment]
@NekR My thoughts are "that's why we have a `dev` branch". New and Production users are generally using `master`.

I would love to give `offline-plugin` 4.5.x some additional exposure (presuming that automated plus manual testing passes). I have relied on automated plus manual testing for dependency updates and so far no major issues.

I will certainly review issues on `offline-plugin` repo before including 4.5.

Thoughts anyone else?

**Comment by NekR** [comment]
@gihrig 
> My thoughts are "that's why we have a dev branch". New and Production users are generally using master.

Yeah, makes sense.

> I will certainly review issues on offline-plugin repo before including 4.5.

There is actually no known bugs in there. I fixed those which I found during PWA development already. But sure, feel free to take a look, maybe you'll find something useful there :-)

**Comment by mxstbr** [comment]
Sounds good to me @gihrig!

**Comment by gihrig** [comment]
@NekR @mxstbr @all `offline-pugin 4.5.2` presented no problems in testing, so I included it in the recent dependency update #1301.

Thanks @NekR ✨

**Comment by NekR** [comment]
@gihrig that's great! Thanks for trying it 👍

**Comment by gretzky** [comment]
Closing due to inactivity, looks like this should be all set
